libbeat/publisher/queue/qos.go

```diff
@@ -0,0 +1,112 @@
+package queue
+
+import (
+	"container/heap"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+type Qoser interface {
+	// Release the process lock and give the processing time to the next available producer
+	ReleaseToken()
+
+	// Schedule picks the next producer who gets to produce to the queue
+	Schedule()
+
+	// Clone an instance of QoSer
+	CreateClient() QoserClient
+}
+
+type QoserClient interface {
+	// Add the producer in the wait queue based
+	RequestToken(*ProducerWrapper, int, time.Time)
+
+	// Wait until the next available slot
+	Wait()
+
+	// Send a signal to the producer to wake up
+	Awaken()
+}
```

Comment:
a beat.Client must be able to disconnect at anytime... you missing a `Close()` as well?